File: BullsAndCows/BullsAndCows.cpp
/*
Author: Matruprasad Chand
Date: 06/10/2018
Info:
	* This is a console based word guessing game
	* The User will be provided with a few clues and has to guess the word correctly within the given number of tries.
	* For each letter in the right place, the user gets a bull and for each correct character placed in the wrong place the user gets a cow.
*/

#include "pch.h" // Why do you bother me? I don't need a precompiled class that does nothing more than annoy me to the end.
#include <iostream>
#include <string>
#include "FBullCowGame.h"
using namespace std;

/* 
	Substituion of types to make C++ more Unreal like
	There are a few different types of strings in Unreal, out of which I am using FText(https://docs.unrealengine.com/en-us/Programming/UnrealArchitecture/StringHandling/FText) and FString(https://docs.unrealengine.com/en-us/Programming/UnrealArchitecture/StringHandling/FString) in this project. 
	FText is unmutable and is used for string values that won't change.
	FString is mutable and is used for string values that will change during the execution of the program.
	FBullCowGame class uses the FString while the main uses FText.
*/
using FText = string;
using int32 = int32;

constexpr int32 WORD_LENGTH = 9;
FBullCowGame BCGame;

void printIntro();
void printStatus(FText guess);
void playGame();
bool willPlayAgain();
FText getGuess();

int main()
{
	do
	{
		printIntro();
		playGame();
	} while (willPlayAgain());
	cout << endl;
	return 0;
}

void playGame()
{
	int32 temp = BCGame.GetMaxTries();
	while (temp > 0)
	{

		FText guess = getGuess();
		cout << endl;
		EGuessStatus status = BCGame.IsGuessValid(guess);
		switch (status)
		{
		case EGuessStatus::OK:
			printStatus(guess);
			temp--;
			break;
		case EGuessStatus::Not_Isogram:
			break;
		case EGuessStatus::Wrong_Length:
			cout << "Please enter a " << BCGame.GetWordLength() << " letter word.\n";
			break;
		case EGuessStatus::Not_lowercase:
			break;
		default:
			break;
		}
	}
}

/*
This method takes a FText input and only accepts a Yes or a No
Why did I make it so complex? No one knows.
This could have been made far simpler by just checking for Y or by only taking a single character.
*/
bool willPlayAgain()
{
	FText response;
	cout << "\nDo you want to play again? (Y/N)\n";
	getline(cin, response);
	if (response[0] == 'Y' || response[0] == 'y')
		return true;
	else if (response[0] == 'N' || response[0] == 'n')
		return false;
	else
	{
		cout << "\nPlease enter either Yes or No.\n";
		return willPlayAgain();
	}
}

void printIntro()
{
	// system("cls"); is used to clear the console window. This is windows specific and works with the Command prompt. In linux the command and code will be different.
	system("cls");
	BCGame.Reset();
	cout << "\nWelcome to Bulls and Cows\n";
	cout << "\nCan you think of " << BCGame.GetWordLength() << " letter isogram I am thinking of?\n";
}

void printStatus(FText guess)
{
	FBullCowCount count = BCGame.SubmitGuess(guess);
	cout << "Bulls = " << count.bull << " Cows = " << count.cow << endl;
}

FText getGuess()
{
	FText Guess = "";

	// This entire system is utterly unnecessary but cool.
	if (BCGame.GetCurrentTry() == 1)
	{
		cout << "\nEnter your " << BCGame.GetCurrentTry() << "st guess: ";
	}
	else if (BCGame.GetCurrentTry() == 2)
	{
		cout << "\nEnter your " << BCGame.GetCurrentTry() << "nd guess: ";
	}
	else if (BCGame.GetCurrentTry() == 3)
	{
		cout << "\nEnter your " << BCGame.GetCurrentTry() << "rd guess: ";
	}
	else
	{
		cout << "\nEnter your " << BCGame.GetCurrentTry() << "th guess: ";
	}
	getline(cin, Guess); // TODO: Check validity of the guess, and only return valid guesses
	return Guess;
}

File: BCGameCSharp/BullCowCSharp/BullsAndCows.cs
﻿/**
Author: Matruprasad Chand
Date: 11/10/2018
Info:
	* This is a console based word guessing game
	* The User will be provided with a few clues and has to guess the word correctly within the given number of tries.
	* For each letter in the right place, the user gets a bull and for each correct character placed in the wrong place the user gets a cow.
*/
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BullCowCSharp
{
    class BullsAndCows
    {
        public BullsAndCows()
        {
            Reset();
            MaxTries = secretWord.Length - 1;
        }

        public BullsAndCows(int x)
        {
            Reset();
            MaxTries = x;
        }

        public int MaxTries { get; private set; }
        public int CurrentTry { get; private set; }
        public bool IsGameWon { get; private set; }

        private static string secretWord; // This is where the differences might start appearing
        private List<string> wordCol = new List<string> { "brick", "blind", "simple", "sample", "world" };

        //Private methods
        //This shows just how simple things are in C#. Random number generation and assignemnt can be done in a single line while in C++ what mt19937 and what else.
        private void SetSecretWord() { secretWord = wordCol.ElementAt(new Random().Next(0, wordCol.Count)); }

        public void Reset()
        {
            IsGameWon = false;
            SetSecretWord();
            CurrentTry = 1;
        }

        private bool IsIsogram(string guess)
        {
            if (guess.Length < 2)
            {
                return true;
            }
            Dictionary<char, bool> letterSeen = new Dictionary<char, bool>();
            foreach (var letter in guess)
            {
                bool x;
                if (letterSeen.TryGetValue(letter, out x)) // Tries to get the value of the given index. If accessed normally like an array, it will throw exception if the key is not present.
                {
                    return false;
                }
                else
                {
                    letterSeen[letter] = true;
                }
            }
            return true;
        }

        // Public methods
        public int GetWordLength() => secretWord.Length; // Lambda expressions in C# are far simpler to use than in C++. But that might be due to practice and familiarity. 

        // Instead of creating a struct I can use a Tuple<> in C#.
        public Tuple<int, int> SubmitGuess(string guess)
        {
            var bull = 0;
            var cow = 0;
            for (var i = 0; i < GetWordLength(); i++)
            {
                for (var j = 0; j < GetWordLength(); j++)
                {
                    if (guess[i] == secretWord[j])
                    {
                        if (i == j)
                        {
                            bull++;
                        }
                        else
                        {
                            cow++;
                        }
                    }
                }

            }
            if (bull == GetWordLength())
            {
                IsGameWon = true;
            }
            SubmitGuess++;
            return Tuple.Create(bull, cow);
        }

        public GuessStatus IsGuessValid(string guess)
        {
            if (!IsIsogram(guess))
            {
                return GuessStatus.Not_Isogram;
            }
            else if (guess.Length != GetWordLength())
            {
                return GuessStatus.Invalid_Length;
            }
            else
            {
                return GuessStatus.OK;
            }
        }
    }
}
